src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -239,12 +239,42 @@ public static void Copy(Array sourceArray, long sourceIndex, Array destinationAr
             return InternalGetValue(GetFlattenedIndex(new ReadOnlySpan<int>(indices)));
         }
 
-        public object? GetValue(int index)
+        public unsafe object? GetValue(int index)
         {
-            if (Rank != 1)
+            MethodTable* pMethodTable = RuntimeHelpers.GetMethodTable(this);
+
+            // The array rank is 0 for SZ arrays, throw if that's not the case
+            if (pMethodTable->MultiDimensionalArrayRank != 0)
+            {
                 ThrowHelper.ThrowArgumentException(ExceptionResource.Arg_Need1DArray);
+            }
+
+            // Validate the length for all cases
+            if ((uint)index >= NativeLength)
+            {
+                ThrowHelper.ThrowIndexOutOfRangeException();
+            }
+
+            // If the element type is a reference type, we know that reading a value as an object
+            // is always valid. This path can be optimized without a call into the native runtime,
```

Comment:
I do not think it makes sense to be introduce a special path like this here. If you would like to make this method better, you can:

- Rewrite InternalGetValue for CoreCLR in C#. You can use the fully managed NativeAOT implementation as a template.
- Optimize `GetFlattenedIndex` by e.g. creating a specialized version for Rank=1 or making it aggresively inlined.

It will provide similar benefits, in much cleaner way.

Reply:
Slowly getting there. I've done these changes now but I'm pretty sure the code to check whether the element type is a class is not right. Is there a better way to do this that I'm missing? Or, what specific flags should I be using? 🤔

Reply:
You would need to rewrite https://github.com/dotnet/runtime/blob/4f010449b1e9fbf28bbac1d1fff35da2316c3314/src/coreclr/classlibnative/bcltype/arraynative.cpp#L939-L975 pretty much 1:1.

- Get TypeHandle for array element
- Check that it is not TypeDesc
- Check that it is ValueType

This is touching a number of structures that are not mirrored to managed.

Reply:
I've started porting this 1:1, wanted to double check I was on the right track. I'm thinking:
- We should add a `RuntimeHelpers.Box(MethodTable*, ref byte)` helper being an FCall to `MethodTable::Box`, so that we can have the managed `GetValue` here call that in case the element type is not a value type. Is this ok?
- I'm seeing that porting `TypeHandle.GetMethodTable()` ends up requiring porting a whole lot of code from the VM, as the path that handles the case where the handle is a bit involved. Should we port all of that, or can we assume that path is not taken, or should we just add another FCall for that to avoid duplicating all of that code? 🤔

Reply:
> Add a RuntimeHelpers.Box(MethodTable*, ref byte) helper being an FCall to MethodTable::Box, so that we can have the managed GetValue here call that in case the element type is not a value type. Is this ok?

Yes.

> I'm seeing that porting TypeHandle.GetMethodTable() ends up requiring porting a whole lot of code from the VM, as the path that handles the case where the handle is a bit involved.

You should not need to port `TypeHandle.GetMethodTable` for this. The check for the unsupported case is more complex than what it needs to be. It can just check for TypeDesc. Once you know that the type is not a typedesc, you can convert the TypeHandle to `MethodTable*` using a simple cast.
```
if (elementType is TypeDesc)
    Throw(Arg_TypeNotSupported);
MethodTable* pElementType = elementType.AsMethodTable(); // AsMethodTable is a simple cast
```

Reply:
Ah, I see, yeah this looks much simpler if none of the cases where we have a type desc is supported 😄